polkadot/runtime/common/src/impls.rs

```diff
@@ -150,6 +151,27 @@ impl TryConvert<&VersionedLocation, xcm::latest::Location> for VersionedLocation
 	}
 }
 
+/// Adapter for [`Contains`] trait to match [`VersionedLocatableAsset`] type converted to the latest
+/// version of itself where it's location matched by `L` and it's asset id by `A` parameter types.
+pub struct ContainsLatest<L, A>(PhantomData<(L, A)>);
+impl<L, A> Contains<VersionedLocatableAsset> for ContainsLatest<L, A>
```

Comment:
I would move it to the `xcm-builder`, maybe here: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/matches_location.rs ?

Reply:
it's implemented over `VersionedLocatableAsset` type, which is part of common.

Reply:
I see, I thought `VersionedLocatableAsset` was part of the XCM stuff, so maybe we could move everything related to `VersionedLocatableAsset` and `ContainsLatest` to one place in the XCM builder., e.g.: 
https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/locatable_asset.rs.
_(I think I saw some complaints about adding too much stuff to the `polkadot-runtime-common`.)_

As I see, we already have `LocatableAssetId` in `xcm-builder` here: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/pay.rs#L191-L198, which could possibly also go to the `locatable_asset.rs`

And maybe `VersionedLocatableAsset` could also wrap that `LocatableAssetId`, but I think this can be just from V5, e.g.:
```
pub enum VersionedLocatableAsset {
...
	#[codec(index = 4)]
-        V4 { location: xcm::v4::Location, asset_id: xcm::v4::AssetId },
+	V4(LocatableAssetId),
}
```

Reply:
> Since we should be more conservative with xcm-builder...

Where is this coming from? Can you elaborate more?

> Every time someone see these types, ( `VersionedLocatableAsset` and `LocatableAssetId`), they suggest to remove it and use only `Location`, refactor it somehow, etc.

Well, perhaps it's worth considering refactoring it then.


I think keeping `VersionedLocatableAsset` outside of `xcm` / `xcm-builder`, where we currently have all `Versioned*` stuff and also `LocatableAssetId`, seems kind of strange. As far as I can see, it is used by relay runtimes and collectives runtime, so from this perspective, `polkadot-runtime-common` is not the best place (`xcm-builder` seems better):
```
[package]
name = "polkadot-runtime-common"
description = "Pallets and constants used in Relay Chain networks."
```

`VersionedLocatableAsset` contains v3 and v4. How do you propose keeping this up-to-date? Do you expect that when we add `xcm::v5`, someone will also update this? If so, there's no way to know from the xcm module's point of view (considering that one day we might move xcm out of polkadot-sdk) that some versioned stuff is used elsewhere. If not, for example, when we add `xcm::v6`, we remove `xcm:v3`. Is that okay? I think the check-migrations jobs should detect this and fail.

Another thing to consider, I see it is used for `pallet_collective` as the key for `ConversionRateToNative` or `pallet_treasury` as `AssetKind`. I think it's a similar situation to what's discussed here: https://github.com/paritytech/polkadot-sdk/issues/1130. And if I recall correctly, the last time we spoke with Joe about `pallet_assets` and `Location` as `AssetId`, everything pointed to the solution of using `xcm::latest::prelude::Location` as `AssetId` (with migrations) and `VersionedLocation` as `AssetIdParameter` (which is used for extrinsics).

I think `VersionedLocatableAsset` presents a very similar situation. Perhaps it's worth considering splitting `AssetKind` into `AssetKind` / `AssetKindParameter`. For example, this could help avoid using `Box` for extrinsic parameters like `asset_kind: Box<T::AssetKind>`, though I'm not sure if this is good or bad, it can be left for the runtime to decide. My point is that ideally, we should have one method/pattern for storing XCM-related data.

Reply:
> Since we should be more conservative with xcm-builder

I meant between `polkadot-runtime-common` and `xcm-builder`, we should be more conservative with the last. It's better if we keep this as it is, and move later. Then moving it now, and removing it if we decide we do not need it. 

I do not see a problem with having it under this common crate, and using it in system chain, I think we have more cases like this. 
If we add new xcm version, and do not update this type, it wont compile because of the match. But we should not rely on the types location and match, there should be tests catching the issue. And in many cases if this types does not have the latest version, it is not a breaking issue. 

With this PR I would not move anything. I would try to explore first what's the alternatives to this type, since it was discussed few times. If we have another alternative, we introduce it and remove this type. If we keep this type, we can moving it.

Reply:
I agree there's no need to do the refactor here, but I would very much look into the refactor. I've previously had a go at simplifying this in https://github.com/paritytech/polkadot-sdk/pull/2006. This `LocatableAsset` was introduced to know when an asset was outside of the local chain to do a `spend` from the treasury and know to which location to send it. I argue it's enough with just an asset id for the asset, and then we use the location of the beneficiary (or sender) to know where to send it.

Reply:
I will check that draft PR, and I am also looking for a solution where that PayOverXcm can also detect that the asset is located on local chan, but beneficiary is on Asset Hub. We have this use case when treasury wants to fund it's own account on Asset Hub. I will initiate a discussion around it, and open an RFC.
@franciscoaguirre please approve this PR if you happy with it.

Reply:
I only want to understand the purpose of this change. Is it because DOT was being passed on to the asset rate converter? Or is it to allow things like USDT to be used 1:1 when they come from system parachains?

Reply:
Hmm, with sovereign accounts it definitely becomes harder. My draft PR was of the idea that we could just use the beneficiary location to know to which chain to send it to. It wouldn't work when we want to send it somewhere different than the location to fund a sovereign account.

We might need then a location for the chain and another for the beneficiary, like on the `send` extrinsic in pallet-xcm. There's no need to name it locatable asset though, just a regular asset, a regular destination, and a regular beneficiary.

Reply:
> I only want to understand the purpose of this change. Is it because DOT was being passed on to the asset rate converter? Or is it to allow things like USDT to be used 1:1 when they come from system parachains?

The asset rate does not know there are different locations of native asset (as in the example in the PR description). And will require a rate to be set for it. With there types we make asset conversion understand different locations of native asset, and do not require a rate to be set for it.

Reply:
I guess that's one of the problems of marking the difference between the same asset on different chains